File: src/main/java/com/tabishev/leshy/bytecode/Expression.java
package com.tabishev.leshy.bytecode;

import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.Type;

import java.util.function.Consumer;
import java.util.function.Function;

public interface Expression {
    ExpressionKind push(MethodVisitor writer);

    static Expression voidExpression() {
        return new SimpleExpression(ExpressionKind.Void, writer -> {});
    }

    static Expression constExpression(int value) {
        return constExpression(ExpressionKind.Int, value);
    }

    static Expression constExpression(long value) {
        return constExpression(ExpressionKind.Long, value);
    }

    static Expression constExpression(boolean value) {
        if (value) {
            return constExpression(ExpressionKind.Int, 1);
        } else {
            return constExpression(ExpressionKind.Int, 0);
        }
    }

    static Expression negate(Expression arg) {
        return new UnaryExpression(arg, kind -> kind.negateInst.get());
    }

    static Expression sum(Expression arg1, Expression arg2) {
        return new BinaryExpression(arg1, arg2, kind -> kind.sumInst.get());
    }

    static Expression mult(Expression arg1, Expression arg2) {
        return new BinaryExpression(arg1, arg2, kind -> kind.multInst.get());
    }

    static Expression local(int idx, ExpressionKind kind) {
        return new SimpleExpression(kind, writer -> writer.visitVarInsn(kind.loadInst.get(), idx));
    }

    static Expression thisInstance() {
        return new SimpleExpression(ExpressionKind.Object, writer -> writer.visitVarInsn(Opcodes.ALOAD, 0));
    }

    static Expression invokeSuper(Class<?> superClass) {
        return writer -> {
            writer.visitVarInsn(Opcodes.ALOAD, 0);
            writer.visitMethodInsn(Opcodes.INVOKESPECIAL, Type.getInternalName(superClass), "<init>", "()V", false);
            return ExpressionKind.Void;
        };
    }

    static Expression cast(Expression expression, Class<?> clz) {
        return writer -> {
            var kind = expression.push(writer);
            writer.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(clz));
            return kind;
        };
    }

    private static Expression constExpression(ExpressionKind kind, Object value) {
        return new SimpleExpression(kind, writer -> writer.visitLdcInsn(value));
    }
}

// todo: use records for this classes
class UnaryExpression implements Expression {
    final Expression arg;
    final Function<ExpressionKind, Integer> instruction;

    UnaryExpression(Expression arg, Function<ExpressionKind, Integer> instruction) {
        this.arg = arg;
        this.instruction = instruction;
    }

    @Override
    public ExpressionKind push(MethodVisitor writer) {
        var kind = arg.push(writer);
        writer.visitInsn(instruction.apply(kind));
        return kind;
    }
}

class BinaryExpression implements Expression {
    final Expression arg1;
    final Expression arg2;
    final Function<ExpressionKind, Integer> instruction;

    BinaryExpression(Expression arg1, Expression arg2, Function<ExpressionKind, Integer> instruction) {
        this.arg1 = arg1;
        this.arg2 = arg2;
        this.instruction = instruction;
    }

    @Override
    public ExpressionKind push(MethodVisitor writer) {
        var kind1 = arg1.push(writer);
        var kind2 = arg2.push(writer);
        assert kind1 == kind2;
        writer.visitInsn(instruction.apply(kind1));
        return kind1;
    }
}

class SimpleExpression implements Expression {
    final ExpressionKind kind;
    final Consumer<MethodVisitor> push;

    public SimpleExpression(ExpressionKind kind, Consumer<MethodVisitor> push) {
        this.kind = kind;
        this.push = push;
    }

    @Override
    public ExpressionKind push(MethodVisitor writer) {
        push.accept(writer);
        return kind;
    }
}

File: src/main/scala/com/tabishev/leshy/bytecode/Expression.scala
package com.tabishev.leshy.bytecode

import org.objectweb.asm.{MethodVisitor, Opcodes, Type}

import java.lang.reflect.Method
import scala.reflect.ClassTag

trait Expression {
  def push(writer: MethodVisitor): BytecodeExpressionKind
}

object Expression {
  def negate(arg: Expression): Expression =
    UnaryExpression(_.negateInst.get, arg)
  def sum(arg1: Expression, arg2: Expression): Expression =
    BinaryExpression(_.sumInst.get, arg1, arg2)
  def mult(arg1: Expression, arg2: Expression): Expression =
    BinaryExpression(_.multInst.get, arg1, arg2)

  def const(value: Int): Expression = ConstExpression(BytecodeExpressionKind.Int, value)
  def const(value: Long): Expression =  ConstExpression(BytecodeExpressionKind.Long, value)
  def const(value: Boolean): Expression = ConstExpression(BytecodeExpressionKind.Int, if (value) 1 else 0)

  def void(): Expression = VoidExpression

  def invokeVirtual(clazz: Class[_], name: String, args: Expression*): InvokeMethod =
    InvokeMethod(Opcodes.INVOKEVIRTUAL, clazz, name, args.toSeq)

  def invokeStatic(clazz: Class[_], name: String, args: Expression*): InvokeMethod =
    InvokeMethod(Opcodes.INVOKESTATIC, clazz, name, args.toSeq)

  def local[T: ClassTag](idx: Int): Expression = Local(idx, BytecodeExpressionKind.of[T])
}

case class Cast(expression: Expression, clz: Class[_]) extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = {
    val kind = expression.push(writer)
    writer.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(clz))
    kind
  }
}

case object VoidExpression extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = BytecodeExpressionKind.Void
}

case class ConstExpression private[bytecode](kind: BytecodeExpressionKind, value: Any) extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = {
    writer.visitLdcInsn(value)
    kind
  }
}

case class Field(isStatic: Boolean, name: String, owner: Type, tpe: Type) extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = {
    if (!isStatic) {
      ThisInstance().push(writer)
      writer.visitFieldInsn(Opcodes.GETFIELD, owner.getInternalName, name, tpe.getDescriptor)
    } else {
      ???
    }
    BytecodeExpressionKind.of(tpe)
  }
}

case class InvokeMethod(opcode: Int, clazz: Class[_], name: String, args: Seq[Expression]) extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = {
    args.foreach { arg =>
      assert(arg.push(writer) != BytecodeExpressionKind.Void)
    }
    writer.visitMethodInsn(opcode, Type.getType(clazz).getInternalName, name, Type.getMethodDescriptor(method()), false)
    BytecodeExpressionKind.of(Type.getType(method().getReturnType))
  }

  private def method(): Method = {
    val methods = clazz.getMethods.filter(_.getName == name)
    assert(methods.length == 1)
    methods(0)
  }
}

case class Local(idx: Int, kind: BytecodeExpressionKind) extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = {
    writer.visitVarInsn(kind.loadInst.get, idx)
    kind
  }
}

case class ThisInstance() extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = {
    writer.visitVarInsn(Opcodes.ALOAD, 0)
    BytecodeExpressionKind.Object
  }
}

case class InvokeSuper(superClass: Class[_]) extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = {
    writer.visitVarInsn(Opcodes.ALOAD, 0)
    writer.visitMethodInsn(Opcodes.INVOKESPECIAL, Type.getInternalName(superClass), "<init>", "()V", false)
    BytecodeExpressionKind.Void
  }
}

case class BinaryExpression(opcode: BytecodeExpressionKind => Int, arg1: Expression, arg2: Expression) extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = {
    val kind = arg1.push(writer)
    val kind2 = arg2.push(writer)
    assert(kind == kind2)
    writer.visitInsn(opcode(kind))
    kind
  }
}

case class UnaryExpression(op: BytecodeExpressionKind => Int, arg: Expression) extends Expression {
  override def push(writer: MethodVisitor): BytecodeExpressionKind = {
    val kind = arg.push(writer)
    writer.visitInsn(op(kind))
    kind
  }
}
